Adds a tooltip to the transaction tables to show the imported payee

matt-fidd opened the pull request:
Putting this up for PR to get some opinions. (Fixes #782)

I think a way to view the imported payee (other than creating a rule from a transaction) would be hugely beneficial and, to me, this is cleaner than another column in the table.

I'm not sure if it would be obvious to people what it was showing so maybe it needs an `Imported Payee: ${payeeName}` in the tooltip instead.

Anyway, let me have your thoughts please! 😄 

To test, import a budget file with some imported transactions, then hover over the payee field in the table.

Based on #2828 so a lot of the diff is unrelated.

youngcw commented:
How does one view this tooltip?

matt-fidd commented:
> How does one view this tooltip?

Sorry, just updated the PR details. 
To test, import a budget file with some imported transactions, then hover over the payee field in the table.

youngcw commented:
I would say that the tooltip should be bigger, and it probably would help to add in "imported payee" to it, like you were saying.  The delay to show could maybe be a touch longer too, but thats not a big deal right now

matt-fidd commented:
Upped the delay to 750ms which feels a bit better, and added a bit of padding.

I've got two options for the layout:
1:
![image](https://github.com/actualbudget/actual/assets/81489167/517bfe1e-1eef-4d58-bcdb-d98f550578a4)

2:
![image](https://github.com/actualbudget/actual/assets/81489167/4819703a-9a5b-464b-8eb0-1acfc4a58a4d)

There's also the option for tooltip placement, I've changed it to pop out from the right instead of the top but I'm not sure if I prefer it.

youngcw commented:
I prefer the second option.

youngcw commented:
Would it be worth adding in some sort of indicator that there is an imported_payee that can be shown?

matt-fidd commented:
It could be useful, but I don't want to clutter the table. Do you have any suggestions as to how it might look?

youngcw commented:
I agree that if there is anything it should be really simple and not intrusive.  Maybe like a subdued icon of some kind.  But that also could be for a later PR.

psybers commented:
What about a dotted underline?  Usually dotted underlines mean there is more info/hovers available. Solid underlines mean link.

Though I think the dotted underlines usually require you to still click them to get the hover and it is not automatic.  Still might work in this case?

psybers commented:
Can we avoid showing the imported payee if it is identical to the current payee?

psybers commented:
I also wonder if the tooltip placement would not be better under the payee name, instead of to the right of it.  That would allow for more easily comparing the two.

psybers commented:
https://ux.stackexchange.com/questions/133212/standard-style-for-text-that-opens-hover-tool-tip

Something like this? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/abbr#providing_an_expansion

psybers commented on upcoming-release-notes/2828.md:
Not sure what happened here, but you seem to have two PRs in here? There are unrelated changes from #2828 in here.

matt-fidd replied:
I mentioned this in my comments, #2828 migrated this file to the new Tooltip/Popover react-aria so it's based on top of that branch. Once it's merged in I'll rebase to master

matt-fidd commented:
> I also wonder if the tooltip placement would not be better under the payee name, instead of to the right of it. That would allow for more easily comparing the two.

You're right, done

matt-fidd commented:
> What about a dotted underline? Usually dotted underlines mean there is more info/hovers available. Solid underlines mean link.
> 
> Though I think the dotted underlines usually require you to still click them to get the hover and it is not automatic. Still might work in this case?

Good idea! That's subtle enough and doesn't make it look too busy. Try with your own file if you can to get a better idea

![image](https://github.com/actualbudget/actual/assets/81489167/cd3bbca8-b1a6-4d63-ba1f-4b5329a2734b)

matt-fidd commented:
> Can we avoid showing the imported payee if it is identical to the current payee?

I'm not sure on this, I personally prefer it always-on.

psybers commented:
It just seems odd to see things like this:

<img width="266" alt="image" src="https://github.com/actualbudget/actual/assets/1115390/e00efcaf-c7ab-4335-8897-8baab5683af2">

You could easily do this by changing the `importedPayee ?` guard to be `importedPayee && payee != importedPayee ?`.

youngcw commented:
I feel like seeing that they are the same could be useful.  Like "these are the same.  Maybe I should make a real payee for this"

MatissJanis commented:
(updated the base branch so it's easier to see the real diff)

matt-fidd commented:
Yes, you're probably both right. I think it made less sense to me because I have very few importedPayees the same as payee.

Updated, let me know what you think

matt-fidd commented:
The netlify build appears to be struggling

youngcw commented:
If the original problem this PR was trying to solve was to make the imported payee visible in the gui, doesn't selectively not showing it kinda defeat the whole point?  Like there is no way to know if a transaction actually has an imported payee since there could be one, but its just not showing.  There is no way to know without resorting to the original workarounds we were trying to not need.

psybers commented:
@youngcw What problem(s) was being solved with this PR? The original issue was to be able to view the imported payee.  Even with selectively applying the tooltips, you are viewing the imported payee (because it is showing in the payee column).  If you then change the payee to something else, it would show the imported payee.

What use case(s) are enabled by not selectively filtering?

youngcw commented:
Im thinking about the use case of trying to make some rules based on imported payee, but you want to see what that imported payee is first on a few transactions (like similar but slightly different imported payee).  

It looks like, as of now, that if the imported payee and the payee name match, then there is no indication that an imported payee exists at all.  You would have to know before hand that the imported payee and the payee happen to be matching. (Often its obvious because of weird formatting, but thats not always the case)  So if you didn't know beforehand that they were the same, and see that the payee doesn't have the option of a tooltip, the expected meaning of that would be that there is no imported payee.

Another side use case would be if you wanted to know if a transaction came from import or manual addition, the indicator of imported payee would be a quick visual of that.

psybers commented:
> > The original issue was to be able to view the imported payee
> 
> I agree that this is the goal: to be able to view the imported payee. And wouldn't hiding the fact that there is one defeat this goal?

I guess that is all in how you interpret the semantics of the request.  I interpret it as being able to view the imported payee.  Which for these cases, is already being displayed in the payee column. Hence, it is already fulfilled for those cases and we should focus on the others.

> Another side use case would be if you wanted to know if a transaction came from import or manual addition, the indicator of imported payee would be a quick visual of that.

Glad you mentioned this second case!  I think we need indicators that show if it was manually entered, imported via bank sync, or added via API.

youngcw commented:
@matt-fidd I don't know why this specific PR is having issues with netlify.  The others seem fine.

matt-fidd commented:
I wonder if the netlify issue is because the workflow isn't being run because of the different base branch. I can see that it might just target merges into `master`.

youngcw commented:
> I wonder if the netlify issue is because the workflow isn't being run because of the different base branch. I can see that it might just target merges into `master`.

Lets find out.  Can you try pushing some dummy commit?

matt-fidd commented:
Yes, that's sorted it.

matt-fidd commented:
With #2828 merged, this is back open for a proper review.

psybers commented on packages/desktop-client/src/components/transactions/TransactionsTable.jsx:
```diff
@@ -562,17 +564,46 @@ function PayeeCell({
+            placement="bottom"
+            triggerProps={{ delay: 750 }}
+          >
+            <div
+              style={{ display: 'flex', alignItems: 'center', width: '100%' }}
+            >
+              <PayeeIcons
+                transaction={transaction}
+                transferAccount={transferAcct}
+                onNavigateToTransferAccount={onNavigateToTransferAccount}
+                onNavigateToSchedule={onNavigateToSchedule}
+              />
+              <UnexposedCellContent
+                {...props}
+                style={{ borderBottom: `1px dashed ${theme.pageTextSubdued}` }}
+              />
+            </div>
```
Is it possible to pull this up into it's own variable, to avoid repeating it below?  The only difference is the `style` on the `UnexposedCellContent`, which you could easily set with a ternary.

Then the code would look more like `importedPayee ? <tooltip>$var</tooltip> : $var`.

matt-fidd replied:
I'm not ignoring this, it _can't_ be hard but I appear to be fighting with React. Will give it another go with a fresh head tomorrow. I think it's something props related.

matt-fidd replied:
Ha! Got it! Updated now.

matt-fidd commented:
Seem to be having some trouble with split transactions and upcoming schedules, the flex grow
![image](https://github.com/user-attachments/assets/1d254984-5df4-40bc-8141-d834f5aa8884)

is causing the border to extend
![image](https://github.com/user-attachments/assets/df0a87e7-3940-419a-af59-f19e56e53197)

matt-fidd commented:
It looks like the transactions that are behaving have an extra div from somewhere. (.css-83enl9 in this case)

![image](https://github.com/user-attachments/assets/2e197693-0e34-4dcb-8936-5c5971e7daf7)

Does anyone know how I can track this down? I can't see any obvious difference between a child and a normal transaction. Also happening with scheduled transactions that have not been posted.

matt-fidd commented:
In VRT, they're all coming up full width.

![image](https://github.com/user-attachments/assets/7914c4b1-c17b-4409-b619-cfec88187e5f)

youngcw commented:
I wonder if the vrt issues is because of a smaller screen size?

psybers commented on packages/desktop-client/src/components/transactions/TransactionsTable.jsx:
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
+            />
+            <UnexposedCellContent
+              {...props}
+              style={{ borderBottom: `1px dashed ${theme.pageTextSubdued}` }}
```
Shouldn't this only be underlined if there is an importedPayee?

matt-fidd replied:
https://github.com/actualbudget/actual/pull/3018#issuecomment-2222878509

matt-fidd replied:
Sorry, as I was eating I only just realised what you meant. Will fix.

psybers commented on packages/desktop-client/src/components/transactions/TransactionsTable.jsx:
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
+            content={
+              <View style={{ padding: 10 }}>
+                <Text style={{ fontWeight: 'bold' }}>Imported Payee</Text>
+                <Text style={{ fontWeight: 'normal' }}>{importedPayee}</Text>
```
Is the normal needed here, shouldn't that already be the case?
```suggestion
                <Text>{importedPayee}</Text>
```

matt-fidd replied:
Yes, think this left over from a refactor where I had them all in the same line and needed to reset.

psybers commented on packages/desktop-client/src/components/transactions/TransactionsTable.jsx:
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
-      )}
+      unexposedContent={props => {
+        const payeeInfo = (
+          <div style={{ display: 'flex', alignItems: 'center', width: '100%' }}>
```
Is this the VRT issue?  You set the width to 100% here so the children will fill it?

matt-fidd replied:
Most likely, I have no idea why I did that... must have been a reason. Will check when I can get back to a dev box.

matt-fidd commented:
I'll update VRT when I get home after work.

youngcw commented:
Im seeing inconsistent lengths of the underline.  Some full box some right length.  Im also seeing that many of the schedule icons are getting misaligned and appear above the payee, but not all of them weirdly.

matt-fidd commented:
Netlify mustn't have updated for me properly, it looked fine a few minutes ago but I can see the same issues now.
That must have been the reason for the full width flex.

youngcw commented:
Looking good!  I noticed that hovering the transfer or schedule icons will show the tooltip too.  Thats not a big problem but it does seem a little weird.

psybers commented:
> Looking good! I noticed that hovering the transfer or schedule icons will show the tooltip too. Thats not a big problem but it does seem a little weird.

Probably my fault for suggesting lifting the common code into a single place. Now the tooltip wraps everything...

matt-fidd commented:
Fixed now, I've moved the icons out of the tooltip

matt-fidd commented:
Updated to fix conflicts from recent parent payee display change.

youngcw commented:
If the transaction is a split then there doesn't seem to be a way to view the imported payee there.  Am I missing something?

matt-fidd commented:
It looks like the imported payee information isn't present in a child transaction.

![image](https://github.com/user-attachments/assets/bfb515fc-eec1-4484-ba77-3b8a1f2a714d)

I'll have to rewrite it to pull in the info from the parent.

matt-fidd commented:
Alternatively, should the imported payee be pulled through to the child on creation as it is with payee? I imagine there's a migration that could be written to pull the info through historically if we go in that direction.

youngcw commented:
Could we still have the option to hover and view it on the parent?  The child payees might be different, plus the child payees weren't the ones that got imported, most of the time.

matt-fidd commented:
> Could we still have the option to hover and view it on the parent? The child payees might be different, plus the child lines weren't the ones that got imported, most of the time.

Done, it looks a little different because of the split button it's inside but works the same.

matt-fidd commented:
Do you think the delay is a bit long or about right?

youngcw commented:
The delay time feels good to me
